File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.py

def solution(bandage, health, attacks):
    maxhealth=health        #최대 체력
    time=0                  #현재 시간
    maxtime=attacks[-1][0]   #마지막 공격 시간
    success=0               #연속성공

    attackDic={}
    for i in attacks:
        attackDic[i[0]]=i[1]    # {1:15, 5:16, 8:6}

    while time<=maxtime:
        if time in attackDic:   #몬스터 공격 타이밍
            health-=attackDic[time]
            success=0

            if health<=0:
                return -1
        else:
            success+=1
            if success==bandage[0]:  # 추가회복
                health+=bandage[1]+bandage[2]
                success=0

                if(health>maxhealth):
                    health=maxhealth
                
            else:
                health+=bandage[1]
                if(health>maxhealth):
                    health=maxhealth

        time+=1

            
    return health

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.java
class Solution {
    public int solution(int[] bandage, int health, int[][] attacks) {
        int limit=attacks[attacks.length-1][0]+1;   // 12
        int[] healths=new int[limit];
        int t=bandage[0];   // 붕대 감기 시간
        int x=bandage[1];   // 초당 회복량
        int y=bandage[2];   // 추가 회복량
        int seq=0;          // 붕대 연속 감기 시간
        healths[0]=health;
        int i=1;
        for(int j=0;j<attacks.length;j++){
            int time=attacks[j][0];     // 공격 시점
            int damage=attacks[j][1];   // 데미지
            while(i<limit){
                if(i==time){    // 쳐맞으면
                    seq=0;  // 붕대 초기화
                    healths[i]=healths[i-1]-damage;     // 피깎고
                    if(healths[i]<=0){   // 뒤짐
                        return -1;
                    }
                    i++;
                    break;
                }else{
                    seq++;
                    healths[i]=healths[i-1]+x;  // 회복
                    if(seq==t){     // 추가회복
                        healths[i]=healths[i]+y;
                        seq=0;  // 초기화
                    }
                    if(healths[i]>health){
                        healths[i]=health;
                    }

                    i++;
                }
            }
        }
        return healths[limit-1];

//        int answer = 0;
//        return answer;
    }
}